File: Data Structures & Algorithms/string-encode-and-decode/submission-3.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        ret = ''

        for s in strs:
            ret += str(len(s)) + '#' + s
        
        return ret

    def decode(self, s: str) -> List[str]:
        ret = []
        i = 0
        while i < len(s):
            k = i + 1
            while s[k] != '#':
                k += 1
            length = int(s[i:k])

            j = k + 1
            ret.append(s[j:j + length])
            i = j + length

        return ret

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    productExceptSelf(nums) {
        let prefix = 1
        let postfix = 1
        const ret = new Array(nums.length).fill(1)

        for (let i = 0; i < nums.length; i++) {
            ret[i] = prefix
            prefix *= nums[i]
        }

        for (let i = nums.length - 1; i >= 0; i--) {
            ret[i] *= postfix
            postfix *= nums[i]
        }

        return ret
    }
}
